File: src/main/java/tmi/recipe/types/RecipeItem.java
package tmi.recipe.types;

import arc.graphics.g2d.TextureRegion;

import java.util.Objects;

public abstract class RecipeItem<T> implements Comparable<RecipeItem<?>>{
  public final T item;

  protected RecipeItem(T item) {
    this.item = item;
  }

  public abstract int ordinal();
  public abstract int typeID();
  public abstract String name();
  public abstract String localizedName();
  public abstract TextureRegion icon();
  public abstract boolean hidden();

  public boolean hasDetails(){ return false; }
  public void displayDetails(){}

  public boolean locked() {
    return false;
  }

  @Override
  public int compareTo(RecipeItem<?> o) {
    int n = Integer.compare(typeID(), o.typeID());

    if (n == 0){
      return ordinal() - o.ordinal();
    }

    return n;
  }

  @Override
  public boolean equals(Object object) {
    if (this == object) return true;
    if (object == null || getClass() != object.getClass()) return false;
    RecipeItem<?> that = (RecipeItem<?>) object;
    return Objects.equals(item, that.item);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name());
  }
}

File: src/main/kotlin/tmi/recipe/types/RecipeItem.kt
package tmi.recipe.types

import arc.Core
import arc.graphics.g2d.TextureRegion

abstract class RecipeItem<T> protected constructor(@JvmField val item: T) : Comparable<RecipeItem<*>> {
  abstract val ordinal: Int
  abstract val typeOrdinal: Int
  abstract val typeID: Int
  abstract val name: String
  abstract val localizedName: String
  abstract val icon: TextureRegion
  abstract val hidden: Boolean
  abstract val hasDetails: Boolean
  abstract val locked: Boolean

  open fun displayDetails() {}

  override fun compareTo(other: RecipeItem<*>): Int {
    val n = typeOrdinal.compareTo(other.typeOrdinal)

    if (n == 0) {
      return ordinal - other.ordinal
    }

    return n
  }

  override fun toString(): String {
    return "($item)"
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as RecipeItem<*>
    return item == that.item
  }

  override fun hashCode(): Int {
    return typeID*31 + name.hashCode()
  }
}
